### PR TITLE
Support Mastodon-style actor profile attachments

### PR DESCRIPTION
**Pull request by relistan**
This PR adds support for Mastodon-style attachments on actors, which are most noticeably visible when viewing the profile of an actor. This supports, like Mastodon, up to 4 attachments on a single actor, and adds to the Settings form a few fields to make this editable. I am not a web UI expert so I can't say that this is the best way to do it, but it works well. It might be nice to start with one field and be able to click something to add more fields in the future. For now it just displays 4 empty fields when nothing is present.

This has been tested running on my own Ktistec instance, and seems to do all the right things. The way the JSON post works to the settings controller is exactly the same as the HTML, which, while great for URL-encoded forms, is a little weird in JSON. Otherwise I think this is all a pretty good fit for how things are implemented.
Screen Shot 2022-12-09 at 12 54 33 PM

### CODE REVIEW

File: src/assets/css/main.less [hunk 1]
```diff
@@ -195,13 +202,16 @@ body { display: flex; flex-direction: column; }
 
 /* missing styling for forms */
 .ui.form { margin-bottom: 1rem; }
+#half { width: calc(48% - 1ex); margin-right: 1ex; }
 .ui.form:last-child { margin-bottom: 0; }
 .ui.inline.form { display: inline-block; }
```
Comment (relistan) [inline src/assets/css/main.less:207]: Tried many, many things and cannot get the styles to apply to elements in the form to limit their width (ran webpack, etc).

File: src/controllers/settings.cr [hunk 2]
```diff
@@ -78,7 +79,15 @@ class SettingsController
...
-      "site" => params["site"]?.try(&.to_s.presence)
+      "site" => params["site"]?.try(&.to_s.presence),
+
+      "attachments" => 1.upto(4).reduce(Hash(String, String).new) do |memo, i|
+        if params["attachment_#{i}_name"]?.try(&.to_s.presence) &&
+            params["attachment_#{i}_value"]?.try(&.to_s.presence)
+          memo[params["attachment_#{i}_name"].to_s] = params["attachment_#{i}_value"].to_s
+        end
+        memo
+      end.to_json
```
Comment (relistan) [inline src/controllers/settings.cr:90]: Will break out into a method

File: src/views/settings/settings.html.slang [hunk 3]
```diff
@@ -10,6 +10,23 @@ p
...
+	p
+	  | You can have up to 4 items displayed on your profile.
+	div#half class="field"
+	  label 
+	    | Name
+	div#half class="field"
+	  label
+	    | Value
+    - 0.upto(3).each do |i|
+      - if actor.attachments[i].nil?
```
Comment (relistan) [inline src/views/settings/settings.html.slang:24]: Not sure how, with everything I did, that this nil check is still required.

File: src/views/partials/actor-panel.html.slang [hunk 4]
```diff
@@ -17,6 +17,13 @@
       br: a href=actor.display_link = actor.account_uri
       - if (summary = actor.summary.presence)
         br: == s summary
+	.ui.segment
+	  - if (attachments = actor.attachments.presence)
+        table
+          - attachments.each do |attachment|
+            tr
+              th.name == attachment.name
+              td.value == maybe_show_link(attachment.value)
```
Comment (relistan) [inline src/views/partials/actor-panel.html.slang:26]: Wanted to add a helper here to display links as HTML in the HTML and JSON views, so we can store them as plain links. But was not able to figure out where to put it that it would be in scope.

**Comment by toddsundsted** [comment]
@relistan thank you! i'll check this out and try it in the morning!

you should consider adding support to map the fields from activitypub messages from other servers—there are a few steps necessary, like updating the json-ld context of the message, and mapping fields, but interoperability would be a huge win, and is easier up front than retrofitting. tests would be great, as well—let me know if i can help you navigate the existing tests!

**Comment by relistan** [comment]
@toddsundsted thanks 🙏, yeah definitely tests will happen once I know how the main code works 😁. Good point about mapping from other servers. I'll see how that works in mastodon. The main objective for me is to make the Ktistec actors look better when viewed by others who might be interested in following.

**Comment by toddsundsted** [comment]
activitypub support isn't a ton of extra code—it's usually the interoperability testing that's the pain, depending on how many other servers support a feature and how many of their edge cases and incompatibilities you run into. it looks like pleroma, for example, decided to at least support viewing mastodon's profile metadata (https://git.pleroma.social/pleroma/pleroma/-/issues/810) in additional to their own solution, and i can't convince myself that other servers (pixelfed, etc.) support metadata in a federated way, at all.

that said, if you don't want to wade into that immediately, i'm happy to bite that off. in practice that would mean that i'd take responsibility for the model and the activitypub wiring.

i'm also happy to just answer questions. let me know how i can help!

**Comment by relistan** [comment]
@toddsundsted Thanks for looking things over. I'm happy to do that, just will take me a couple of days to find more time. Since we live in Ireland and Thanksgiving was not a public holiday, we are doing the big day today.

File: src/framework/framework.cr [hunk 5]
```diff
@@ -61,9 +61,9 @@ module Ktistec
     end
 
     def assign(options)
-      @host = options["host"] if options.has_key?("host")
-      @site = options["site"] if options.has_key?("site")
-      @footer = options["footer"] if options.has_key?("footer")
+      @host = options["host"].to_s if options.has_key?("host")
+      @site = options["site"].to_s if options.has_key?("site")
+      @footer = options["footer"].to_s if options.has_key?("footer")
```
Comment (relistan) [inline src/framework/framework.cr:66]: This was only working before because the values were all `String`s

File: src/framework/model.cr [hunk 6]
```diff
@@ -365,6 +365,8 @@ module Ktistec
               if (o = options[key]).is_a?(typeof(self.{{v}}))
                 @changed << {{v.symbolize}}
                 self.{{v}} = o.as(typeof(self.{{v}}))
+              else
+                Log.warn { "Got invalid type assigned to #{key}" }
```
Comment (relistan) [inline src/framework/model.cr:369]: This would have really helped identify the issue with saving, so leaving it in.

Comment (toddsundsted) [reply src/framework/model.cr:369]: this needs to go into the initialize methods and the other assign method, but it's probably overdue. i'd recommend you pull the change out of your PR and i'll make all of the necessary changes in commit it myself. keeping a dirty copy of this file around until then is probably a good idea—i've been bitten myself before...

File: src/framework/framework.cr (hunk 5 above)
Comment (toddsundsted) [reply src/framework/framework.cr:66]: this is a good catch. i'd recommend `.as(String)` instead, which will ensure it has the right type but won't accidentally coerce some non-string value into a string.

Comment (relistan) [reply src/framework/framework.cr:66]: Good call

**Comment by relistan** [comment]
@toddsundsted I was  not worrying about safety of the content in the field since this is a single user server. But actually, thinking about this, if we do support fields in federation from other servers, we will have to strip all HTML except links. I went through the PR on Pleroma implementing this and that's what they do also: https://git.pleroma.social/pleroma/pleroma/-/merge_requests/1488/diffs#b3fe63dfecb918b7b8a768711068cd8a8cbadbea_83_88

**Comment by toddsundsted** [comment]
without looking at the code, it looks like these profile metadata items are either text (no markup), or html anchor tags with a boatload of attributes. i'm assuming they are transforming text that looks like a url into one of these anchors. in any case, it probably makes sense to assume the contents are html and to sanitize it on output. this is what i do for summary, for example.  there's already a sanitizer (named s) which you can call from a template: e.g. `== s summary`

File: src/views/partials/actor-panel.html.slang (hunk 4 above)
Comment (toddsundsted) [reply src/views/partials/actor-panel.html.slang]: converting url strings to html links would be a useful utility function. it could go into something like src/framework/util.cr

Comment (toddsundsted) [reply src/views/partials/actor-panel.html.slang]: the other angle is maybe punt on this entirely for the time being. if the values can be html then anyone can author links in html. otherwise, it seems like you have to either convert urls to links when the values are first set, or you have to do it on the fly when they are presented. if the former, then users will have to deal with a URL they entered presented back to them as html when they edit it. if the later, then the link creation will have to be smart enough to, at least, not make urls inside of the href attribute of an existing link into a link and breaking the html.

imo, the right long term solution is probably to implement rich text/html editing for these fields, as well as summary and site fields like the footer. then users can just wysiwyg their profile metadata, summary, etc. etc.

Comment (relistan) [reply src/views/partials/actor-panel.html.slang]: The implementation that is in this PR currently has them written and stored as plaintext to make editing easy, and converted to HTML links on presentation. Since Mastodon converts to HTML in the JSON as well, I also did that. If you were to enter HTML in the original input field, it will just return it back without any processing, because the link detector is intentionally only looking for strings that _start_ with what appears to be a proper URL. If you were to put your own HTML link in there, you would get that back unmodified next time you edit. With regard to federation, I feel like that's most easily solved with sanitizing on intake.

Comment (toddsundsted) [reply src/views/partials/actor-panel.html.slang]: 👍 

one downside to sanitizing on intake is that you lose local access to the original content. (the upside is that you do the sanitizing once.) i'm a data nerd so that's always less attractive.

Comment (relistan) [reply src/views/partials/actor-panel.html.slang]: Sounds good. Either way works, if you'd prefer to keep the original content, I will do that. I don't think this is a high volume activity since it would only happen when you view federated profile on Ktistec, so doing it on render seems OK. Just means that we'll always present sanitized HTML, even if it was created locally.

Comment (toddsundsted) [reply src/views/partials/actor-panel.html.slang]: you can have it both ways. for objects we have content and media_type, which are activitypub fields, and a source, which is the original (with a media type, as well, in case we support markdown, etc. in the future.

(i'm not arguing for this, btw—i don't know if i'd go to the trouble, myself, but i'm pointing it out in case you find it valuable)

**Comment by relistan** [comment]
I rebased on `main` and built a copy and deployed to my site. It *almost* does the right thing. I seems that links have to be in the very specific format that Mastodon wants, or they just show up as text. Will test with Pleroma and Pixelfed (if it supports attachments). If that all works, I'll write the tests.

Screen Shot 2022-11-28 at 1 57 13 PM

**Comment by toddsundsted** [comment]
this looks great! i'm surprised mastodon let you get away without including the necessary definitions in the @context. (i'd have assumed it would just ignore them without it.)

**Comment by relistan** [comment]
Thanks! AFAICT it doesn't have any additional context entries, I had checked that out.

**Comment by relistan** [comment]
Was prepared to write some tests today but can't get the specs to even finish running. I start them, some stuff gets emitted, and then it just hangs there forever.

File: spec/models/activity_pub/actor_spec.cr [hunk 7]
```diff
@@ -295,26 +297,32 @@ Spectator.describe ActivityPub::Actor do
 
     it "adds a public following relationship" do
       foo_bar.follow(other, confirmed: true, visible: true).save
-      expect(foo_bar.all_following(public: true)).to eq([other])
-      expect(foo_bar.all_following(public: false)).to eq([other])
+      expect(foo_bar.all_following(public: true).size).to eq(1)
+      expect(foo_bar.all_following(public: true).first.iri).to eq(other.iri)
+      expect(foo_bar.all_following(public: false).size).to eq(1)
+      expect(foo_bar.all_following(public: false).first.iri).to eq(other.iri)
```
Comment (relistan) [inline spec/models/activity_pub/actor_spec.cr:303]: A bunch of these tests were testing exact matches on the entire object. This was failing because the DB has a default value of JSON `[]` for attachments, so once it was saved it was `attachments = []` rather than `attachments = nil`. Converted to check that the result has 1 item and that the `iri`s match.

File: README.md [hunk 8]
```diff
@@ -190,6 +190,15 @@ Once these steps are done, you're running!
 
 <img src="https://raw.githubusercontent.com/toddsundsted/ktistec/main/images/o0ton2.png" width=640>
 
+## Running Tests
+
+The the tests requires that you set the `KEMAL_ENV` environment variable before
+running them, or they will lock up. Running the tests is:
+
+```
+KEMAL_ENV=test crystal spec
```
[the remark this answers is not in the document]
Comment (relistan) [reply README.md:199]: No worries, I eventually figured it out.

File: src/framework/json_ld.cr [hunk 9]
```diff
@@ -104,7 +104,7 @@ module Ktistec
     end
 
     private def self.expand_value(term, value, context, loader)
-      if term.in?(["https://www.w3.org/ns/activitystreams#content", "https://www.w3.org/ns/activitystreams#name", "https://www.w3.org/ns/activitystreams#summary"])
+      if term.in?(["https://www.w3.org/ns/activitystreams#content", "https://www.w3.org/ns/activitystreams#name", "https://www.w3.org/ns/activitystreams#summary", "https://www.w3.org/ns/activitystreams#attachment"])
```
Comment (relistan) [inline src/framework/json_ld.cr:107]: @toddsundsted I'm not 100% sure with this change. It required changing how `ActivityPub::Object` mapped from `ld_json` to the properties. I will note this below in comments. But this was necessary to get the attachments to work for `Actor` AFAICT.

File: src/models/activity_pub/actor.cr [hunk 10]
```diff
@@ -733,10 +771,28 @@ module ActivityPub
         summary: dig?(json, "https://www.w3.org/ns/activitystreams#summary", "und"),
         icon: dig_id?(json, "https://www.w3.org/ns/activitystreams#icon", "https://www.w3.org/ns/activitystreams#url"),
         image: dig_id?(json, "https://www.w3.org/ns/activitystreams#image", "https://www.w3.org/ns/activitystreams#url"),
-        urls: dig_ids?(json, "https://www.w3.org/ns/activitystreams#url")
+        urls: dig_ids?(json, "https://www.w3.org/ns/activitystreams#url"),
+        attachments: attachments_from_ldjson(
+          json.dig?("https://www.w3.org/ns/activitystreams#attachment")
+        )
```
Comment (relistan) [inline src/models/activity_pub/actor.cr:777]: This is why the change above to `json_ld.cr` was necessary.

File: src/models/activity_pub/object.cr [hunk 11]
```diff
@@ -606,8 +606,8 @@ module ActivityPub
           Tag::Mention.new(name: name, href: href)
         end,
         attachments: dig_values?(json, "https://www.w3.org/ns/activitystreams#attachment") do |attachment|
-          url = attachment.dig?("https://www.w3.org/ns/activitystreams#url").try(&.as_s)
-          media_type = attachment.dig?("https://www.w3.org/ns/activitystreams#mediaType").try(&.as_s)
+          url = attachment["url"].try(&.as_s)
+          media_type = attachment["mediaType"].try(&.as_s)
```
Comment (relistan) [inline src/models/activity_pub/object.cr:610]: These changes were necessitated by the change to `ld_json.cr`. The two things labeled attachment don't have the same schema. I think the Mastodon implementation on `actor` is non-standard. The activitiystreams context has
```
    "attachment": {
      "@id": "as:attachment",
      "@type": "@id"
    },
```

That doesn't match what Mastodon uses for actors.

If there is a better way to implement this, let me know. This is what I could figure out.

**Comment by relistan** [comment]
Rebased on `main`

**Comment by relistan** [comment]
@toddsundsted I think this is ready to go!

**Comment by toddsundsted** [comment]
@relistan thanks again for all of this effort!!

i'm looking at the json-ld support for the handling of object and actor attachments. this _should be_ possible with json-ld and shouldn't require changes to json-ld here or to the mapping of objects. but... 1) i'm not 100% sure of that so i want to doublecheck, 2) i'm not 100% sure my implementation of json-ld handles nested properties correctly so i want to look at that, 3) i'm wondering how this works in the wild.

my only other suggestion is to relax the sql definition of the new "attachments" column for actor to permit nulls. you are already handling `nil` checks for attachments in all the right places in the code. and there are several changes to specs caused by failing object-level comparison that defaulting to "[]" is causing. i'm not arguing that this is the best way to define the schema, but it would be better to minimize changes in this PR and then go back in a separate PR to fix that if there's a better way to do it.

otherwise this looks great! i'd love to find a way to get rid of the view helper and maybe just fix up attachments in a before save filter in the actor model, but i didn't see any obvious/quick way to do that, and i'm generally fine with code that's not perfectly dry, so i wouldn't hold up merging this on account of that.

**Comment by toddsundsted** [comment]
only minimal changes are required to handle attachments, without changes to the json-ld code or the object model. the missing piece is some more context, specifically:
```
-          "https://w3id.org/security/v1"
+          "https://w3id.org/security/v1",
+          {
+            "schema":"http://schema.org#",
+            "PropertyValue":"schema:PropertyValue",
+            "value":"schema:value"
+          }
```
in a few places. these lines ensure that the type and the value are mapped correctly. it looks like mastodon provides them in json for objects and ktistec depends on the context to properly interpret keys in a json object. i'll add some comments where the necessary changes are. (note: mastodon provides these, but it looks like they ignore them and don't expand json-ld documents using the context. til.)

File: spec/models/activity_pub/actor_spec.cr [hunk 12]
```diff
@@ -166,7 +171,12 @@ Spectator.describe ActivityPub::Actor do
           "mediaType": "image/jpeg",
           "url": "image link"
         },
-        "url":"url link"
+        "url":"url link",
```
Comment (toddsundsted) [inline spec/models/activity_pub/actor_spec.cr:174]: the `@context` section in this json-ld document needs to be changed to:
```
        "@context":[
          "https://www.w3.org/ns/activitystreams",
          "https://w3id.org/security/v1",
          {
            "schema":"http://schema.org#",
            "PropertyValue":"schema:PropertyValue",
            "value":"schema:value"
          }
        ],
...
```
this will ensure that when the json-ld is expanded, the attachment type and value are expanded correctly. name is already defined by the activity streams core schema.

File: spec/models/activity_pub/actor_spec.cr [hunk 13]
```diff
@@ -191,6 +201,24 @@ Spectator.describe ActivityPub::Actor do
       expect(actor.icon).to eq("icon link")
       expect(actor.image).to eq("image link")
       expect(actor.urls).to eq(["url link"])
+
+      expect(actor.attachments).not_to be_nil
+      attachments = actor.attachments.not_nil!
+      expect(attachments.size).to eq(2)
+      expect(attachments.all? { |a| a.type == "PropertyValue" }).to be_true
```
Comment (toddsundsted) [inline spec/models/activity_pub/actor_spec.cr:208]: once the attachment type is expanded, the value here will be "http://schema.org#PropertyValue" instead of just "PropertyValue"

File: spec/models/activity_pub/actor_spec.cr [hunk 14]
```diff
@@ -223,6 +251,20 @@ Spectator.describe ActivityPub::Actor do
       expect(actor.icon).to eq("icon link")
       expect(actor.image).to eq("image link")
       expect(actor.urls).to eq(["url link"])
+
+      expect(actor.attachments.not_nil!.size).to eq(2)
+      expect(actor.attachments.not_nil!.all? { |a| a.type == "PropertyValue" }).to be_true
```
Comment (toddsundsted) [inline spec/models/activity_pub/actor_spec.cr:256]: this should be changed to "http://schema.org#PropertyValue"

File: spec/models/activity_pub/actor_spec.cr (hunk 7 above)
Comment (toddsundsted) [reply spec/models/activity_pub/actor_spec.cr:352]: if the column is defined to accept null (see my longer reply to the PR) then these changes won't be necessary.

File: spec/models/tag/hashtag_spec.cr [hunk 15]
```diff
@@ -12,7 +12,7 @@ Spectator.describe Tag::Hashtag do
 
     it "associates with an actor" do
       expect{described_class.new(subject: actor, name: "actor").save}.to change{Tag::Hashtag.count}.by(1)
-      expect(described_class.find(name: "actor").subject).to eq(actor)
+      expect(described_class.find(name: "actor").subject.iri).to eq(actor.iri)
```
Comment (toddsundsted) [inline spec/models/tag/hashtag_spec.cr:15]: if the column is defined to accept null (see my longer reply to the PR) then these changes won't be necessary.

File: src/framework/json_ld.cr (hunk 9 above)
Comment (toddsundsted) [reply src/framework/json_ld.cr:107]: this shouldn't be necessary once some additional definitions are added to the @context.

File: src/models/activity_pub/actor.cr [hunk 16]
```diff
@@ -733,10 +788,28 @@ module ActivityPub
...
         "image" => dig_id?(json, "https://www.w3.org/ns/activitystreams#image", "https://www.w3.org/ns/activitystreams#url"),
-        "urls" => dig_ids?(json, "https://www.w3.org/ns/activitystreams#url")
+        "urls" => dig_ids?(json, "https://www.w3.org/ns/activitystreams#url"),
+        "attachments" => attachments_from_ldjson(
+          json.dig?("https://www.w3.org/ns/activitystreams#attachment")
+        )
       }.compact
     end
 
+    def self.attachments_from_ldjson(entry)
```
Comment (toddsundsted) [inline src/models/activity_pub/actor.cr:798]: once changes to @context are in place, the keys here will be in their expanded form. e.g.
```
    def self.attachments_from_ldjson(entry)
      entry_not_nil = (entry.try(&.as_a) || [] of JSON::Any).not_nil!

      entry_not_nil.reduce([] of Attachment) do |memo, a|
        name = dig?(a, "https://www.w3.org/ns/activitystreams#name", "und")
        type = a.dig?("@type").try(&.as_s)
        value = dig?(a, "http://schema.org#value")

        unless name.nil? || name.empty? || type.nil? || type.empty? || value.nil? || value.empty?
          value = maybe_wrap_link(value)
          memo << Attachment.new(name, type, value)
        end
        memo
      end
    end
```
"type" becomes "@type" and it's value will be "http://schema.org#PropertyValue"

File: src/models/activity_pub/object.cr (hunk 11 above)
Comment (toddsundsted) [reply src/models/activity_pub/object.cr:610]: this definition is okay. it's defining attachment (both for actors and objects) as the activity streams ("as") attachment. the missing piece is the definition of the PropertyValue type and value. if you make the other changes i suggest, you should be able to get rid of this change.

File: src/views/actors/actor.json.ecr [hunk 17]
```diff
@@ -50,6 +50,9 @@
     "url":<%= image.inspect %>
   },
   <%- end -%>
+  <%- if (attachments = actor.attachments) -%>
```
Comment (toddsundsted) [inline src/views/actors/actor.json.ecr:53]: this is fine, but at the top of this file you need to extend the context:
```
-    "https://w3id.org/security/v1"
+    "https://w3id.org/security/v1",
+    {
+      "schema":"http://schema.org#",
+      "PropertyValue":"schema:PropertyValue",
+      "value":"schema:value"
+    }
```
this will ensure that any json-ld clients (ktistec or others) will know how to expand and handle PropertyValue and value.

**Comment by toddsundsted** [comment]
i'm pretty sure i've attached comments with all of the changes i made to get the json-ld stuff working. if not, let me know! i'm happy to take another look. i tested this with mastodon accounts, and they work. it does not work with your ktistec implementation because it's expecting some information in the context about how to expand PropertyValue and value.

File: spec/models/activity_pub/actor_spec.cr (hunk 12 above)
Comment (relistan) [reply spec/models/activity_pub/actor_spec.cr:174]: Great, thanks! Will try to get it cleaned up this weekend

File: spec/models/activity_pub/actor_spec.cr (hunk 7 above)
Comment (relistan) [reply spec/models/activity_pub/actor_spec.cr:352]: Yeah, I really wanted to not have to accept null, but as I recall, the framework wouldn't compile with that defined as `not null`, and then there is the whole issue where you can't set a default value in the DB and have it respected. I will look at backing out that change, although since I have it running this way on my site it will involve some gymnastics to back it out. A default value of empty array is much better than null, but it seems to be swimming upstream.

File: spec/models/tag/hashtag_spec.cr (hunk 15 above)
Comment (relistan) [reply spec/models/tag/hashtag_spec.cr:15]: I suggest that checking whole object equality here is not a good idea in general.

File: src/models/activity_pub/actor.cr (hunk 16 above)
Comment (relistan) [reply src/models/activity_pub/actor.cr:798]: @toddsundsted I'm obviously doing something wrong. I added the schema info to the jsonld in the test 
```
  let(json) do
    <<-JSON
      {
        "@context":[
          "https://www.w3.org/ns/activitystreams",
          "https://w3id.org/security/v1",
          {
            "schema":"http://schema.org#",
            "PropertyValue": "schema:PropertyValue",
            "value": "schema:value"
          }
        ],
...
```


and used the code above. But I am not getting data that looks like that. Inside the loop in `self.attachments_from_ldjson` I'm still getting e.g.:

```
{"name" => "Blog", "type" => "PropertyValue", "value" => "https://somewhere.example.com"}
```

Comment (toddsundsted) [reply src/models/activity_pub/actor.cr:798]: ok, i'll take a look. there are probably a couple changes that have to be made at the same time. in the meantime, i pushed the changes that i made to [a branch](https://github.com/toddsundsted/ktistec/tree/add-attachments) on github so you could see what i did when i was working on this. this worked against mastodon (but not against your build because of the missing @context information).

relevant commit: https://github.com/toddsundsted/ktistec/commit/24a63f0fe28162f15ff1f6f0250207dcdc7c1d91

File: spec/models/activity_pub/actor_spec.cr (hunk 7 above)
Comment (toddsundsted) [reply spec/models/activity_pub/actor_spec.cr:352]: removing it shouldn't break your site and existing records. your code already handles `nil` in the right places, and loading an empty array from existing records should just work—it looked like you already reduced the case of four empty slots in the array to an empty array (if not it would be easy to handle) so that one edge case should be okay, as well.

i'm not opposed to the default being empty array, btw. i'm just trying to keep the surface area of changes small so that i can more easily reason about them.

File: spec/models/tag/hashtag_spec.cr (hunk 15 above)
Comment (toddsundsted) [reply spec/models/tag/hashtag_spec.cr:15]: i'm open to the argument. semantically it should be fine, however, and equality comparisons at the object level are supported and meant to be correct.

File: src/models/activity_pub/actor.cr (hunk 16 above)
Comment (toddsundsted) [reply src/models/activity_pub/actor.cr:798]: (it's possible you have to undo the change to json_ld.cr as well)

Comment (relistan) [reply src/models/activity_pub/actor.cr:798]: 🤦 Missed that I had not removed that change to `json_ld.cr`. The ActivityPub stuff is all new to me.

File: spec/models/activity_pub/actor_spec.cr (hunk 7 above)
Comment (relistan) [reply spec/models/activity_pub/actor_spec.cr:357]: Sorry, just meant that I have to manually copy the table data over to a new table and rename back, because sqlite doesn't support live schema modifications other than additive operations.

**Comment by relistan** [comment]
@toddsundsted sorry, at this point I think this has taken you as much time as if you had just written it. I *think* it's good enough to ship now. I have not tested it live on my site yet but can do that if you are happy with this implementation now.

**Comment by toddsundsted** [comment]
thanks @relistan! it has taken me a few days to do the review. to clean up the history, i squashed your commits and created another PR: https://github.com/toddsundsted/ktistec/pull/60

i'm going to close this on.
